controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +292,117 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
 	}()
 
 	rebuildingAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeRebuildingController, volName)
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		vbc.logger.Infof("Volume %v is faulted, skip rebuilding", volName)
+		delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		return nil
+	}
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	isOfflineRebuildEnabled, err := vbc.isVolumeOfflineRebuildEnabled(vol.Spec.OfflineRebuilding)
+	if err != nil {
+		return err
+	}
+
+	if isOfflineRebuildEnabled && !util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachmentForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol)
+		if err != nil {
+			return err
+		}
+		if isVolumeInRebuilding {
+			return nil
+		}
+	}
 
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuilding longhorn.VolumeOfflineRebuilding) (bool, error) {
+	if offlineRebuilding == longhorn.VolumeOfflineRebuildingEnabled {
+		return true, nil
+	}
+
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return offlineReplicaRebuildingValue && offlineRebuilding != longhorn.VolumeOfflineRebuildingDisabled, nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachmentForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	areAllReplicasHealthy, err := vbc.isVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas)
+	if err != nil {
+		return va, err
+	}
+	if !areAllReplicasHealthy {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.AnyValue, longhorn.AttacherTypeVolumeRebuildingController)
+	}
+	return va, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas int, replicas map[string]*longhorn.Replica) (bool, error) {
+	healthyCount := 0
+	for _, replica := range replicas {
+		node, err := vbc.ds.GetNodeRO(replica.Spec.NodeID)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return false, err
```

Comment:
Is the intention is to requeue when failing to get the node? An inline comment here might help clarify, as the different error handling may not be immediately obvious.

Reply:
Yes. Added an inline comment. Please review it again